orchagent/neighorch.cpp

```diff
@@ -579,5 +650,438 @@ bool NeighOrch::removeNeighbor(const NeighborEntry &neighborEntry)
+        return false;
+    }
+
+    vector<FieldValueTuple> attrs;
+
+    FieldValueTuple eiFv ("encap_index", to_string(attr.value.u32));
+    attrs.push_back(eiFv);
+
+    FieldValueTuple macFv ("neigh", mac.to_string());
+    attrs.push_back(macFv);
+
+    string key = alias + ":" + ip_address.to_string();
+    m_tableVoqSystemNeighTable->set(key, attrs);
+
+    return true;
+}
+
+bool NeighOrch::voqSyncDelNeigh(string &alias, IpAddress &ip_address)
+{
+    //Sync only local neigh. Confirm for the local neigh and
+    //get the system port alias for key for syncing to GLOBAL_APP_DB
+    Port port;
+    if(gPortsOrch->getPort(alias, port))
+    {
+        if(port.m_system_port_info.type == SAI_SYSTEM_PORT_TYPE_REMOTE)
+        {
+            return true;
+        }
+        alias = port.m_system_port_info.alias;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Port does not exist for %s!", alias.c_str());
+        return false;
+    }
+
+    string key = alias + ":" + ip_address.to_string();
+    m_tableVoqSystemNeighTable->del(key);
+
+    return true;
+}
+
+bool NeighOrch::addKernelRoute(string odev, IpAddress ip_addr)
+{
+    string cmd, res;
+
+    SWSS_LOG_ENTER();
+
+    string ip_str = ip_addr.to_string();
+
+    if(ip_addr.isV4())
+    {
+        cmd = "ip route add " + ip_str + "/32 dev " + odev;
+        SWSS_LOG_NOTICE("IPv4 Route Add cmd: %s",cmd.c_str());
+    }
+    else
+    {
+        cmd = "ip -6 route add " + ip_str + "/128 dev " + odev;
+        SWSS_LOG_NOTICE("IPv6 Route Add cmd: %s",cmd.c_str());
+    }
+
+    int32_t ret = swss::exec(cmd, res);
+
+    if(ret)
+    {
+        /* Just log error and return */
+        SWSS_LOG_ERROR("Failed to add route for %s, error: %d", ip_str.c_str(), ret);
+        return false;
+    }
+
+    SWSS_LOG_INFO("Added route for %s on device %s", ip_str.c_str(), odev.c_str());
+    return true;
+}
+
+bool NeighOrch::delKernelRoute(IpAddress ip_addr)
+{
+    string cmd, res;
+
+    SWSS_LOG_ENTER();
+
+    string ip_str = ip_addr.to_string();
+
+    if(ip_addr.isV4())
+    {
+        cmd = "ip route del " + ip_str + "/32";
+        SWSS_LOG_NOTICE("IPv4 Route Del cmd: %s",cmd.c_str());
+    }
+    else
+    {
+        cmd = "ip -6 route del " + ip_str + "/128";
+        SWSS_LOG_NOTICE("IPv6 Route Del cmd: %s",cmd.c_str());
+    }
+
+    int32_t ret = swss::exec(cmd, res);
+
+    if(ret)
+    {
+        /* Just log error and return */
+        SWSS_LOG_ERROR("Failed to delete route for %s, error: %d", ip_str.c_str(), ret);
+        return false;
+    }
+
+    SWSS_LOG_INFO("Deleted route for %s", ip_str.c_str());
+    return true;
+}
+
+bool NeighOrch::addKernelNeigh(string odev, IpAddress ip_addr, MacAddress mac_addr)
+{
+    SWSS_LOG_ENTER();
+
+    string cmd, res;
+    string ip_str = ip_addr.to_string();
+    string mac_str = mac_addr.to_string();
+
+    if(ip_addr.isV4())
+    {
+        cmd = "arp -s " + ip_str + " " + mac_str + " -i " + odev;
```

Comment:
Can you explain why using arp cmd for IPV4 neighbor, but ip neigh cmd for IPV6 neighbor? Can we use ip neigh cmd for both?

Reply:
Yes. we'll use "ip neigh" command for both IPv4 and IPv6